File: 01_Arrays_&_Hashing/01_Easy/ValidAnagram.cpp
#include <string>
#include<unordered_map>

using namespace std;

class Solution {
public:
    bool isAnagram(string s, string t){
        sort(s.begin(), s.end());
        sort(t.begin(), t.end());

        return s == t;
    };

    bool isAnagramLinear(string s, string t){
        unordered_map<char, int> sCount;
        unordered_map<char, int> tCount;

        if (s.length() != t.length()) return false;

        for (int i = 0; i < s.size(); i++){
            sCount[s[i]]++;
            tCount[s[i]]++;
        }

        return sCount == tCount;
    };
 
};

File: 01_Arrays_&_Hashing/01_Easy/Python/ValidAnagram.py
'''
Given two strings s and t, return true if t is an 
anagram of s, and false otherwise.

An Anagram is a word or phrase formed by rearranging 
the letters of a different word or phrase, typically 
using all the original letters exactly once.
'''

def isAnagram(s, t):
    return True if sorted(s) == sorted(t) else False 

# If two strings are anagrams, then if we sorted the letters
# of both strings, they should both be equal to each other. 

# Time Complexity: O(n log(n))
# Space Complexity: O(n) since we are not sorting the strings 
#                   and storing a list of the sorted letters


# Another approach which takes linear time is 

from collections import defaultdict
def isAnagramLinear(s, t):
    sCount = defaultdict(int)
    tCount = defaultdict(int)

    if len(s) != len(t):
        return False 

    for i in range(len(s)):
        sCount[s[i]] += 1 
        tCount[t[i]] += 1 
    
    return tCount == sCount 
